stdlib/string.gr

```diff
@@ -519,3 +519,596 @@ export let endsWith = (p: String, s: String) => {
+  let end = ptr + size
+  let mut size = 0n // <- number of UTF-16 code words
+
+  while (ptr < end) {
+    let byte = WasmI32.load8U(ptr, 0n)
+    let n = if ((byte & 0x80n) == 0x00n) {
+      1n
+    } else if ((byte & 0xF0n) == 0xF0n) {
+      4n
+    } else if ((byte & 0xE0n) == 0xE0n) {
+      3n
+    } else {
+      2n
+    }
+    if (n == 4n) {
+      size += 2n
+    } else {
+      size += 1n
+    }
+    ptr += n
+  }
+  // multiply by two to get number of bytes (equivalent to left-shift)
+  tagSimpleNumber(size << 1n)
+}
+
+let encodedLength = (s: String, encoding) => {
+  match(encoding) {
+    UTF32_BE => length(s) * 4,
+    UTF32_LE => length(s) * 4,
+    UTF16_BE => utf16Length(s),
+    UTF16_LE => utf16Length(s),
+    UTF8 => byteLength(s),
+  }
+}
+
+exception MalformedUnicode
+
+@disableGC
+let getCodePoint = (ptr: WasmI32) => {
+  // Algorithm from https://encoding.spec.whatwg.org/#utf-8-decoder
+  let (+) = WasmI32.add
+  let (==) = WasmI32.eq
+  let (>=) = WasmI32.geU
+  let (<=) = WasmI32.leU
+  let (<<) = WasmI32.shl
+  let (&) = WasmI32.and
+  let (|) = WasmI32.or
+
+  let mut codePoint = 0n
+  let mut bytesSeen = 0n
+  let mut bytesNeeded = 0n
+  let mut lowerBoundary = 0x80n
+  let mut upperBoundary = 0xBFn
+
+  let mut offset = 0n
+
+  let mut result = 0n
+
+  while (true) {
+    let byte = WasmI32.load8U(ptr + offset, 0n)
+    offset += 1n
+    if (bytesNeeded == 0n) {
+      if (byte >= 0x00n && byte <= 0x7Fn) {
+        result = byte
+        break
+      } else if (byte >= 0xC2n && byte <= 0xDFn) {
+        bytesNeeded = 1n
+        codePoint = byte & 0x1Fn
+      } else if (byte >= 0xE0n && byte <= 0xEFn) {
+        if (byte == 0xE0n) lowerBoundary = 0xA0n
+        if (byte == 0xEDn) upperBoundary = 0x9Fn
+        bytesNeeded = 2n
+        codePoint = byte & 0xFn
+      } else if (byte >= 0xF0n && byte <= 0xF4n) {
+        if (byte == 0xF0n) lowerBoundary = 0x90n
+        if (byte == 0xF4n) upperBoundary = 0x8Fn
+        bytesNeeded = 3n
+        codePoint = byte & 0x7n
+      } else {
+        throw MalformedUnicode
+      }
+      continue
+    }
+    if (!(lowerBoundary <= byte && byte <= upperBoundary)) {
+      throw MalformedUnicode
+    }
+    lowerBoundary = 0x80n
+    upperBoundary = 0xBFn
+    codePoint = (codePoint << 6n) | (byte & 0x3Fn)
+    bytesSeen += 1n
+    if (bytesSeen == bytesNeeded) {
+      result = codePoint
+      break
+    }
+  }
+  result: WasmI32
+}
+
+let _BYTES_OFFSET = 8n;
+
+// Encodes the given string using the given encoding scheme
+// @param s: String - The input string
+// @param encoding: Encoding - The encoding to use
+// @param includeBom: Bool - Whether to include the byte-order marker in the encoded output
+// @returns Bytes
+@disableGC
+export let encode = (s: String, encoding: Encoding, includeBom: Bool) => {
+  let (>>>) = WasmI32.shrU
+  let (-) = WasmI32.sub
+  let (&) = WasmI32.and
+  let (<) = WasmI32.ltU
+  let (<=) = WasmI32.leU
+  let (==) = WasmI32.eq
```

Comment:
Could we move them down next to the add?

Reply:
Fixed.